File: Olympiad/COCI/official/2007/regional_solutions/bard.pas

(*
  Croatian Regional Competition in Informatics 2007
  Task BARD
  Programming language Pascal
*)

program bard;

const
   MAXN = 100;
   MAXV = 50;

var
   n, v, i, j, k          : integer;
   npjesama, veceras, tko : integer;
   zna                    : array[1..MAXN, 1..MAXV] of boolean;
   prisutan               : array[1..MAXN] of boolean;
   pjevase                : array[1..MAXV] of boolean;
   znasve                 : boolean;

begin
   npjesama := 0;
   readln(n);
   readln(v);

   for i:=1 to n do
      for j:=1 to v do
         zna[i, j] := false;

   for i:=1 to v do begin
      for j:=1 to n do
         prisutan[j] := false;
      for j:=1 to npjesama do
         pjevase[j] := false;

      read(veceras);
      for j:=1 to veceras do begin
         read(tko);
         prisutan[tko] := true;
      end;
      readln;

      if prisutan[1] then begin
         npjesama := npjesama + 1;
         pjevase[npjesama] := true;
      end else begin
         for j:=1 to npjesama do
            for k:=1 to n do
               if prisutan[k] and zna[k, j] then
                  pjevase[j] := true;
      end;

      for k:=1 to n do
         for j:=1 to npjesama do
            if prisutan[k] and pjevase[j] then
               zna[k, j] := true;
   end;

   for i:=1 to n do begin
      znasve := true;
      for j:=1 to npjesama do
         if not zna[i, j] then
            znasve := false;
      if znasve then writeln(i);
   end;
end.

File: Olympiad/COCI/official/2007/regional_solutions/bard.c

/*
  Croatian Regional Competition in Informatics 2007
  Task BARD
  Programming language C
*/

#include <stdio.h>
#include <string.h>

#define MAXN 100
#define MAXV 50

int main()
{
   int N, V, i;
   int npjesama = 0;
   int zna[MAXN+1][MAXV];

   scanf( "%d", &N );
   scanf( "%d", &V );

   for ( i=0; i<V; ++i ) {
      int j, k;
      int veceras;
      int prisutan[MAXN+1];
      int pjevase[MAXV];

      for ( j=1; j<=N; ++j )
         prisutan[j] = 0;
      for ( j=0; j<npjesama; ++j )
         pjevase[j] = 0;

      scanf( "%d", &veceras );
      for ( j=0; j<veceras; ++j ) {
         int tko;
         scanf( "%d", &tko );
         prisutan[tko] = 1;
      }

      if ( prisutan[1] ) {
         pjevase[npjesama++] = 1;
      }
      else {
         for ( j=0; j<npjesama; ++j )
            for ( k=1; k<=N; ++k )
               pjevase[j] |= prisutan[k] && zna[k][j];
      }

      for ( k=1; k<=N; ++k )
         for ( j=0; j<npjesama; ++j )
            zna[k][j] |= prisutan[k] && pjevase[j];
   }

   for ( i=1; i<=N; ++i ) {
      int j, znasve = 1;
      for ( j=0; j<npjesama; ++j )
         znasve &= zna[i][j];
      if ( znasve )
         printf( "%d\n", i );
   }

   return 0;
}
